### PR TITLE
fix(table): mark clicks as 'Invalid' when clicking on empty spaces

### PR DESCRIPTION
**Pull request by C47D**
Fixes one part of #7132 

Before this change when clicking in a table empty space (this happens when the table size is bigger than the total size of the rows and columns) the returned columns and rows were the total table cols and rows plus 1.
With this change we return an invalid click, so the click is 'ignored'.

This is my test case, taken from the `lv_example_table_1` example:
```c
    lv_obj_t * table = lv_table_create(lv_screen_active());

    /*Fill the first column*/
    lv_table_set_cell_value(table, 0, 0, "Name");
    lv_table_set_cell_value(table, 1, 0, "Apple");
    lv_table_set_cell_value(table, 2, 0, "Banana");

    /*Fill the second column*/
    lv_table_set_cell_value(table, 0, 1, "Price");
    lv_table_set_cell_value(table, 1, 1, "$7");
    lv_table_set_cell_value(table, 2, 1, "$4");

    /*Set a smaller height to the table. It'll make it scrollable*/
    lv_obj_set_height(table, 400); /* Make the table bigger */
    lv_obj_center(table);
```
And I've added this log to the table event handler to print out the selected cell:
```patch
diff --git a/src/widgets/table/lv_table.c b/src/widgets/table/lv_table.c
index b1553506d..d5caa99be 100644
--- a/src/widgets/table/lv_table.c
+++ b/src/widgets/table/lv_table.c
@@ -582,6 +582,7 @@ static void lv_table_event(const lv_obj_class_t * class_p, lv_event_t * e)
         lv_result_t pr_res = get_pressed_cell(obj, &row, &col);
 
         if(pr_res == LV_RESULT_OK && (table->col_act != col || table->row_act != row)) {
+            LV_LOG_USER("Row: %d, Col: %d", row, col);
             table->col_act = col;
             table->row_act = row;
             lv_obj_invalidate(obj);
```

Then click into the empty space and print the rows and cols only when clicking inside them, not when clicking in the empty space.

This video is before the change:
[Screencast from 27-10-24 23:26:16.webm](https://github.com/user-attachments/assets/3410acfb-c5fd-47af-85e2-70411914dcae)

This other video is with the change:
[Screencast from 27-10-24 23:25:25.webm](https://github.com/user-attachments/assets/566e69d7-6a0c-4d3e-81db-29f4f6bacc4d)

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: src/widgets/table/lv_table.c [hunk 1]
```diff
@@ -997,6 +997,36 @@ static lv_result_t get_pressed_cell(lv_obj_t * obj, uint32_t * row, uint32_t * c
...
+        if(row) *row = LV_TABLE_CELL_NONE;
+        return LV_RESULT_INVALID;
+    }
+    /* Clicked outside columns on empty space */
+    else if(p.x > (total_column_width + area.x1)) {
+        if(col) *col = LV_TABLE_CELL_NONE;
+        if(row) *row = LV_TABLE_CELL_NONE;
+        return LV_RESULT_INVALID;
+    }
+    else { /*Nothing to do*/ }
```
Comment (kissa96) [inline src/widgets/table/lv_table.c]: Do we really need empty code? Logic wise it is understandable and gets optimized out anyways, but is questionable. So far I've not seen empty branches in LVGL. Could you please remove it?

File: src/widgets/table/lv_table.c [hunk 2]
```diff
@@ -997,6 +997,36 @@ static lv_result_t get_pressed_cell(lv_obj_t * obj, uint32_t * row, uint32_t * c
...
+    int32_t total_row_height = 0;
+    int32_t total_column_width = 0;
+
+    for(idx = 0; idx < table->row_cnt; ++idx) {
+        total_row_height += *(table->row_h);
+    }
+
+    for(idx = 0; idx < table->col_cnt; ++idx) {
+        total_column_width += *(table->col_w);
+    }
```
Comment (liamHowatt) [inline src/widgets/table/lv_table.c]: Did you mean to do this?

```suggestion
    for(idx = 0; idx < table->row_cnt; ++idx) {
        total_row_height += table->row_h[idx];
    }

    for(idx = 0; idx < table->col_cnt; ++idx) {
        total_column_width += table->col_w[idx];
    }
```

File: src/widgets/table/lv_table.c [hunk 3]
```diff
@@ -997,6 +997,36 @@ static lv_result_t get_pressed_cell(lv_obj_t * obj, uint32_t * row, uint32_t * c
     lv_point_t p;
     lv_indev_get_point(lv_indev_active(), &p);
 
+    /* Calculate total height of rows so we know if the click was outside them */
+    lv_area_t area;
+    lv_obj_get_coords(obj, &area);
+
+    uint8_t idx = 0;
```
Comment (liamHowatt) [inline src/widgets/table/lv_table.c]: I can imagine a user with a table with more than 255 rows or columns.

```suggestion
    uint32_t idx = 0;
```

**Review by liamHowatt — COMMENTED** [review]
Thanks for finding and fixing this bug :slightly_smiling_face:

File: src/widgets/table/lv_table.c (hunk 3 above)
Comment (C47D) [reply src/widgets/table/lv_table.c]: Done in [c5db141](https://github.com/lvgl/lvgl/pull/7153/commits/c5db1415dfd875328df3294dd818eabb9107aaa9)

File: src/widgets/table/lv_table.c (hunk 2 above)
Comment (C47D) [reply src/widgets/table/lv_table.c]: Done in [c5db141](https://github.com/lvgl/lvgl/pull/7153/commits/c5db1415dfd875328df3294dd818eabb9107aaa9)

File: src/widgets/table/lv_table.c (hunk 1 above)
Comment (C47D) [reply src/widgets/table/lv_table.c]: Done in [c5db141](https://github.com/lvgl/lvgl/pull/7153/commits/c5db1415dfd875328df3294dd818eabb9107aaa9)

**Review by liamHowatt — DISMISSED** [review]

**Comment by C47D** [comment]
Hi @liamHowatt , I was playing with the fix and noticed we could re-use the already existing logic to detect the clicked cell, I've done the refactor in 6f307e2 and I think it's more readable. Here's a video of the fix:

[Screencast from 28-10-24 23:56:36.webm](https://github.com/user-attachments/assets/5024edb8-d5fb-4a08-bf86-b0d1981f2525)

**Comment by kisvegabor** [comment]
cc @liamHowatt

**Review by liamHowatt — COMMENTED** [review]
I am really sorry for the delay.

Thanks for noticing the duplicated logic and simplifying the fix. It looks good to me!

> I also noticed the top-left cell gets colored in the first click, is that expected? I think it also happens before this patch.

Aha, I figured out that `act_col` and `act_row` start as `0` but they should start as `LV_TABLE_CELL_NONE`. Please add the following to `lv_table_constructor` and that problem goes away:

```c
    table->row_act = LV_TABLE_CELL_NONE;
    table->col_act = LV_TABLE_CELL_NONE;
```

Other than that, the fix looks great to me.

**Comment by C47D** [comment]
Not stalled, waiting for feedback

**Comment by liamHowatt** [comment]
@C47D please see my latest comment https://github.com/lvgl/lvgl/pull/7153#pullrequestreview-2454099530

Everything was good! Just initialize the uninitialized values I mentioned in the constructor to make the strange behavior you observed go away.

**Comment by C47D** [comment]
Hi @liamHowatt , sorry I didn't saw your comment, only the bot one, I'll update the code with your suggestion :smile:

**Comment by liamHowatt** [comment]
@lhdjply  Do you know what the problem could be?

I tried this but the test fails for a different reason.

```diff
diff --git a/tests/src/test_cases/test_file_explorer.c b/tests/src/test_cases/test_file_explorer.c
index e81991326..236543bd5 100644
--- a/tests/src/test_cases/test_file_explorer.c
+++ b/tests/src/test_cases/test_file_explorer.c
@@ -9,14 +9,14 @@
 static lv_obj_t * active_screen = NULL;
 static lv_obj_t * file_explorer_obj;
 static lv_file_explorer_t * file_explorer;
-static lv_table_t * file_table;
+static lv_obj_t * file_table;
 
 void setUp(void)
 {
     active_screen = lv_screen_active();
     file_explorer_obj = lv_file_explorer_create(active_screen);
     file_explorer = (lv_file_explorer_t *)file_explorer_obj;
-    file_table = (lv_table_t *)file_explorer->file_table;
+    file_table = file_explorer->file_table;
 }
 
 void tearDown(void)
@@ -39,57 +39,51 @@ void test_file_explorer_read_dir(void)
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    for(uint8_t i = 0; i < file_table->row_cnt; i++) {
-        if(lv_strcmp(lv_table_get_cell_value(file_explorer->file_table, i, 0), LV_SYMBOL_LEFT "  Back") == 0) {
+    for(uint8_t i = 0; i < lv_table_get_row_count(file_table); i++) {
+        if(lv_strcmp(lv_table_get_cell_value(file_table, i, 0), LV_SYMBOL_LEFT "  Back") == 0) {
             back_row = i;
         }
-        if(lv_strcmp(lv_table_get_cell_value(file_explorer->file_table, i, 0), LV_SYMBOL_DIRECTORY "  dev") == 0) {
+        if(lv_strcmp(lv_table_get_cell_value(file_table, i, 0), LV_SYMBOL_DIRECTORY "  dev") == 0) {
             dev_row = i;
         }
-        if(lv_strcmp(lv_table_get_cell_value(file_explorer->file_table, i, 0), LV_SYMBOL_DIRECTORY "  home") == 0) {
+        if(lv_strcmp(lv_table_get_cell_value(file_table, i, 0), LV_SYMBOL_DIRECTORY "  home") == 0) {
             home_row = i;
         }
     }
 
-    file_table->row_act = dev_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, dev_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/dev/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    for(uint8_t i = 0; i < file_table->row_cnt; i++) {
-        if(lv_strcmp(lv_table_get_cell_value(file_explorer->file_table, i, 0), LV_SYMBOL_DIRECTORY "  shm") == 0) {
+    for(uint8_t i = 0; i < lv_table_get_row_count(file_table); i++) {
+        if(lv_strcmp(lv_table_get_cell_value(file_table, i, 0), LV_SYMBOL_DIRECTORY "  shm") == 0) {
             shm_row = i;
         }
     }
 
-    file_table->row_act = shm_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, shm_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/dev/shm/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    file_table->row_act = back_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, back_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/dev/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    file_table->row_act = back_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, back_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    file_table->row_act = home_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, home_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/home/",
                              lv_file_explorer_get_current_path(file_explorer_obj));
 
-    for(uint8_t i = 0; i < file_table->row_cnt; i++) {
-        if(lv_strcmp(lv_table_get_cell_value(file_explorer->file_table, i, 0), LV_SYMBOL_DIRECTORY "  web_user") == 0) {
+    for(uint8_t i = 0; i < lv_table_get_row_count(file_table); i++) {
+        if(lv_strcmp(lv_table_get_cell_value(file_table, i, 0), LV_SYMBOL_DIRECTORY "  web_user") == 0) {
             user_row = i;
         }
     }
 
-    file_table->row_act = user_row;
-    lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
+    lv_table_set_selected_cell(file_table, user_row, 0);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/home/web_user/",
                              lv_file_explorer_get_current_path(file_explorer_obj));

```

**Comment by lhdjply** [comment]
@liamHowatt    

```diff
diff --git a/tests/src/test_cases/test_file_explorer.c b/tests/src/test_cases/test_file_explorer.c
index e81991326..75ca7202b 100644
--- a/tests/src/test_cases/test_file_explorer.c
+++ b/tests/src/test_cases/test_file_explorer.c
@@ -51,6 +51,7 @@ void test_file_explorer_read_dir(void)
         }
     }
 
+    file_table->col_act = 0;
     file_table->row_act = dev_row;
     lv_obj_send_event(file_explorer->file_table, LV_EVENT_VALUE_CHANGED, NULL);
     TEST_ASSERT_EQUAL_STRING("A:src/test_files/test_file_explorer_folder/dev/",
```

**Comment by C47D** [comment]
Thanks for the explanation @lhdjply , I've pushed the fix.

**Comment by lhdjply** [comment]
Although https://github.com/lvgl/lvgl/pull/7153/commits/2508d131df9307ad743f22602a14273b73a4d27d seems understandable, the issue #7257 cannot be reproduced.

**Comment by C47D** [comment]
> Although [2508d13](https://github.com/lvgl/lvgl/commit/2508d131df9307ad743f22602a14273b73a4d27d) seems understandable, but the issue #7257 cannot be reproduced.

2508d13 was pushed right before I noticed your initial reply, that's why it got reverted, If #7257 can not be reproduced it means it is ok?

**Comment by lhdjply** [comment]
My intention is that the commit https://github.com/lvgl/lvgl/commit/2508d131df9307ad743f22602a14273b73a4d27d cannot reproduce the issue #7257, so, the revert is right.

**Review by liamHowatt — APPROVED** [review]
Thanks @lhdjply for the fix and thanks @C47D for adding an extra comment!

**Review by FASTSHIFT — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]

**Pull request merged**